Python/chapter02/2.4 - Partition/miguel_2.4_soln.py

```diff
@@ -137,23 +151,28 @@ def partition_ll(ll: LinkedList, pivot: int) -> LinkedList:
     elements less than x. They do not need to appear
     in between the left and right partitions.
     Runtime: O(n)
-    Space Complexity: O(n)
+    Space Complexity: O(1)
     :param ll: an input linked list
     :param pivot: a number to partition around
     :return: a linked list that is 'partitioned'
     """
-    left_partition = []  # will contain values < pivot
-    right_partition = []  # will contain values >= pivot
+    left_partition = LinkedList()  # will contain values < pivot
+    right_partition = LinkedList()  # will contain values >= pivot
 
     n = ll.head
     while n is not None:
         if n.data < pivot:
-            left_partition.append(n.data)
+            left_partition.append_to_tail(n)
         else:
-            right_partition.append(n.data)
+            right_partition.append_to_tail(n)
         n = n.next
+    # last element may still be pointing
+    # to earlier elements that are less than
+    # pivot so we need to cut that link
+    right_partition.append_to_tail(None)
     # then, merge left and right partition lists into one linked list
-    return LinkedList(*left_partition, *right_partition)
+    left_partition.tail.next = right_partition.head
```

Comment:
can you add a `def append(self, ll: LinkedList) -> None` to prevent the implementation details from leaking?